nav2_amcl/src/amcl_node.cpp

```diff
@@ -1113,6 +1124,195 @@ AmclNode::initParameters()
+        z_short_ = parameter.as_double();
+        reinit_laser = true;
+      }
+    } else if (param_type == ParameterType::PARAMETER_STRING) {
+      if (param_name == "base_frame_id") {
+        base_frame_id_ = parameter.as_string();
+      } else if (param_name == "global_frame_id") {
+        global_frame_id_ = parameter.as_string();
+      } else if (param_name == "map_topic") {
+        map_topic_ = parameter.as_string();
+      } else if (param_name == "laser_model_type") {
+        sensor_model_type_ = parameter.as_string();
+        reinit_laser = true;
+      } else if (param_name == "odom_frame_id") {
+        odom_frame_id_ = parameter.as_string();
+        reinit_laser = true;
+      } else if (param_name == "scan_topic") {
+        scan_topic_ = parameter.as_string();
+        reinit_laser = true;
+      } else if (param_name == "robot_model_type") {
+        robot_model_type_ = parameter.as_string();
+        reinit_odom = true;
+      }
+    } else if (param_type == ParameterType::PARAMETER_BOOL) {
+      if (param_name == "do_beamskip") {
+        do_beamskip_ = parameter.as_bool();
+        reinit_laser = true;
+      } else if (param_name == "tf_broadcast") {
+        tf_broadcast_ = parameter.as_bool();
+      } else if (param_name == "set_initial_pose") {
+        set_initial_pose_ = parameter.as_bool();
+      } else if (param_name == "always_reset_initial_pose") {
+        always_reset_initial_pose_ = parameter.as_bool();
+      } else if (param_name == "first_map_only") {
+        first_map_only_ = parameter.as_bool();
+      }
+    } else if (param_type == ParameterType::PARAMETER_INTEGER) {
+      if (param_name == "max_beams") {
+        max_beams_ = parameter.as_int();
+        reinit_laser = true;
+      } else if (param_name == "max_particles") {
+        max_particles_ = parameter.as_int();
+        reinit_pf = true;
+      } else if (param_name == "min_particles") {
+        min_particles_ = parameter.as_int();
+        reinit_pf = true;
+      } else if (param_name == "resample_interval") {
+        resample_interval_ = parameter.as_int();
+      }
+    }
+  }
+
+  // Checking if the minimum particles is greater than max_particles_
+  if (min_particles_ > max_particles_) {
+    RCLCPP_WARN(
+      this->get_logger(),
+      "You've set min_particles to be greater than max particles,"
+      "this isn't allowed so they'll be set to be equal.");
+    max_particles_ = min_particles_;
+  }
+
+  // Re-initialize the particle filter
+  if (reinit_pf) {
+    if (pf_ != NULL) {
+      pf_free(pf_);
+      pf_ = NULL;
+    }
+    initParticleFilter();
+  }
+
+  // Re-initialize the odometry
+  if (reinit_odom) {
+    motion_model_.reset();
+    initOdometry();
+  }
+
+  // Re-initialize the lasers and it's filters
+  if (reinit_laser) {
+    lasers_.clear();
+    lasers_update_.clear();
+    frame_to_laser_.clear();
+
+    laser_scan_filter_ = std::make_unique<tf2_ros::MessageFilter<sensor_msgs::msg::LaserScan>>(
```

Comment:
> why not call initMessageFilters like you do with the others?

I did that initially and somehow the node stopped responding.. I'll look into this matter

Reply:
The issue that I had was seen in the terminal: 

`CRITICAL FAILURE: SERVER amcl IS DOWN after not receiving a heartbeat for 4000 ms. Shutting down related nodes.`

Reply:
On further investigation, seems that the issue happens, when the laser scanners subscribers are reset...

Reply:
OK, its also OK if we set some parameters as non-changable. We could make `scan_topic_` a non-dynamic parameter, although it would be a useful one. 

Though, I don't understand why that's happening. Is that reset process taking on the order of seconds? Did you reset these variables beforehand or just calling init (or try both)? I don't think that'll change anything, but it would be diagnostic if it did.

Reply:
> Did you reset these variables beforehand or just calling init (or try both)? I don't think that'll change anything,

Yes, you are right. I tried doing both, but somehow both of them yields the same..

Reply:
Oh, I didn't disconnect the `laser_scan_connection_`. Damn..! Now, I guess it is fine.. let me do a couple of more testing, before the commit..